### DISCUSSION TITLE
Complex API with FastAPI: nested models population, validation against DB, related model retrieval, ... ?

### DISCUSSION THREAD

**Discussion by naquad** [fastapi/fastapi#9290] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
#
```

### Description

Hi all,

I'm having a hard time wrapping my head around FastAPI.

My struggle is mostly around the fact that FastAPI is tightly coupled with Pydantic.

The issues I've encountered so far:

* Combinatorial explosion. A ~class~ ~model~ Pydantic model for *every* variation of the input/output.
* Pydantic can't do async validation. That was a shock.
* Pydantic can't do recursive validation.
* SQLAlchemy / Tortoise / any ORM is not integrated with Pydantic to the extent to provide the required flexibility -> more manual Pydantic management. Example: try to create a Pydantic model from SQLAlchemy model and you can't as any customization results in a manual subclassing. I'm not even talking about reusing constraints and validations.

My small app (7 resources, some are nested) ended up with 22(!) Pydantic models. 7 inputs + 7 outputs (with id & read-only fields like created_at/updated_at, hidden passwords and so forth) + 7 collections (paginated responses) + 1 custom input with postprocessing. At the moment, that machinery can't handle nested resources. Even a simple validation against DB (check if a value exists) has led to manual boilerplate code that tries to resemble model error reporting.

Any try to address the issues leads to the depths of Pydantic and SQLAlchemy. I have a feeling that I'm trying to rewrite a framework rather than work with it, so I came to ask before it is too late.

I'm in confusion here: is that the way to go or am I doing something wrong here? What are the best practices for doing async validation, nested models, custom serialization, ...? Tutorial has shown only Pydantic in all the places but that doesn't work very well so far.

### Operating System

Other

### FastAPI Version

0.94.1

### Python Version

3.10.10

**Comment by yinziyan1206** [comment]
1、What do you mean. Do you need different inputs or outputs in one route? or you can just use dict instead of a model.
2、`async` in python is only used in `io` but not `calculation`. you can use `ProcessPool` but it should not be the bottleneck for performance
3、Pydantic can do recursive validation.
4、There are several ways to change sqlalchemy model to pydantic model like `SQLModel`. And also dataclasses in `SQLAlchemy2.0`. Or you can use `codegen` to create model.

**Comment by naquad** [comment]
1. Including but not limited to. Update != create => update & create inputs may differ (consider PATCH method with literally every attribute optional but no id or r/o attributes). Add here permissions and here comes the combinatorial explosion. If I use plain dicts then I'm losing the OpenAPI spec :( Also, dicts imply that I'm doing the data processing which is more manual boilerplate.
2. I'm not sure how the process/thread pool and validation are related. Could you please elaborate? Also, I'm using async SQLAlchemy and I haven't tried this approach but I already bet that starting another event loop will explode in epic and obscure ways because of the way SQLAlchemy is glued with asyncio or it'll be impossible to test. Plus it'll create another DB session.
3. Properly linked DB records can't be represented. Parent has many Items, Item belongs to Parent, and both have backrefs =>  https://github.com/pydantic/pydantic/issues/4509 . Validation in `BaseModel.from_orm` can't be disabled. My current approach is having a separate hierarchy of Pydantic models for both Parent and Item so that either Item has a Parent or Parent has an Item w/o backref. Those hierarchies are used in the corresponding endpoints so that either Item can expose its Parent w/o going into the loop, or Parent can expose its child Items. But this is an obvious workaround. I'm not sure what to do with that.
4. This is not a way to integrate, but a way to copy-paste in runtime (pydantic_to_sqlalchemy, SQLModel, ...) or statically (codegen). The problem is that Pydantic model and SQLAlchemy (or whatever other DB) model are completely different entities with their own lifecycle and independent state. Hence the issue: when the DB model is being created it allows one set of attributes, when it is being updated it allows another set of attributes; there are other factors such as permissions that influence the allowed attributes.

The whole problem with Pydantic is the absolute lack of dynamism. It is a statically defined class with annotations and there are no means to influence the attribute map in the runtime. At the moment, the only thing I haven't tried out is to dynamically generate the Pydantic model from the SQLAlchemy model (`pydantic.create_model` thing) and during its validation pass the control back to the model gluing it with FastAPI via Depends or something like that. But this is way too much effort and it is easier to abandon the whole thing.

**Comment by Viicos** [comment]
I had the same feeling when coming from Django + Django Rest Framework, where the DRY principle is applied pretty much everywhere. When using FastAPI + SQLAlchemy for the first time, I had a lot of trouble trying to avoid having to define too much models, repeating select statements and handling HTTP 404s for each endpoint, handling relationships without facing recursion errors, etc.

But I don't think FastAPI is to blame here, nor is pydantic. Their primary goal isn't to work with SQLAlchemy or any other ORM, they weren't designed for this at the first place. In fact the only part where this is mentioned is in the docs, along with the `orm_mode` pydantic config option.

I believe this is why `SQLModel` was created, or some [other libraries](https://github.com/tiangolo/pydantic-sqlalchemy). The fact that FastAPI is still before 1.0.0 (as is `starlette`) has to be taken into account as well, and both are actively developed and subject to changes. `SQLAlchemy` also received a major update with  v2.

In the future, we might see a library (perhaps `SQLModel`) making it much easier to work with `FastAPI` and `SQLAlchemy`.

Note: https://github.com/dmontagu/fastapi-utils seems to be offering class-based views, pretty much like DRF.

**Reply by naquad** [reply]
So basically it means that it is that bad and with tons of boilerplate?

**Reply by Viicos** [reply]
I wouldn't say that bad, the approach is just different. You have more control over your API behavior, as you can control every update/create/read model. By I do agree, in most cases this control isn't really necessary, and this is where I find DRF way better (but you'll probably find it harder to handle some edge cases with DRF).

Afaik, this isn't specific to FastAPI, you will encounter the same issues with some other frameworks (e.g. Spring with all the DTOs, repositories and entities).

Having a framework similar to Django, but using these two libraries could be great.
